trigger REMOVE_TOTP event on user credential removal

resah opened the pull request:
Add trigger for REMOVE_TOTP event when user deletes credential via user account REST API (and therefore via account-console)

Closes #15403


Maybe a more general EventType like `REMOVE_CREDENTIAL ` might be needed?

resah committed:
trigger REMOVE_TOTP event on removal of an OTP credential

Closes #15403

Signed-off-by: Theresa Henze <theresa.henze@bare.id>

mposolda reviewed (dismissed):
@resah Thanks for the PR.

Yes, as you pointed, it will be good to rename REMOVE_TOTP event to more generic REMOVE_CREDENTIAL event because this endpoint is not specific only to Totp credentials.

It will be also good to save some metadata about credential to the event details like:
- credential type
- credential id
- credential user label (if present)
Also we should likely make sure that userId is saved with the event as well, so admin is aware who was the user who deleted his credential (maybe it is already present in the event, but will be good to doublecheck).

resah commented:
Hi @mposolda ,

thank you for your fast feedback, I updated the PR to fire a `REMOVE_CREDENTIAL` event with details like credential type, credential id and user label.
 
>   Also we should likely make sure that userId is saved with the event as well, so admin is aware who was the user who deleted his credential (maybe it is already present in the event, but will be good to doublecheck).

The user id is already set by the `AccountRestService` and covered by the test I added.


Now I am uncertain, if the new event type `REMOVE_CREDENTIAL` should also be added/registered elsewhere. I saw that the event type `UPDATE_TOTP` is mentioned in several other resources, like `org.keycloak.services.resources.account.Constants`, some docs, Typescript and JSON files. Please let me know, if this needs more adjustments. Thank you!

tolleiv commented:
Would be great to move forward this this one // cc @mposolda

micbis commented:
Any plans on merging this one?

ahus1 commented:
@resah - unfortunately some changes in main created conflicts for this PR. Please rebase this PR. Thanks!

ahus1 commented:
@resah - I had now a deeper look into these changes, and also the original issue. 

The original issue reports that no email is sent out if a credential is removed. As mentioned in your previous comment, there are additional mentions of `UPDATE_TOTP` in other places. Relevant for sending out the email is the `EmailEventListenerProviderFactory` which listens for different events, and the dispatches different templates. 

Creating a new event will have a knock-on effect that this logic needs to be revisited, documentation need to be updated, and a migration guide for the new event needs to be written.

As the primary reason for this PR is to fix the problem at hand, I suggest to narrow down the scope to fixing the bug only. 

I suggest the following: 

* Stick to the original event `REMOVE_TOTP` to avoid rewriting logic in the email event handling
* In `AccountCredntialResource#removeCredential()` check if the credential is of type `OTPCredentialModel.TYPE`, and then create the event as before in `AccountFormService`:
   `event.event(EventType.REMOVE_TOTP).client(auth.getClient()).user(auth.getUser()).success();`
* Optional: Add additional details that seem reasonable (like `SELECTED_CREDENTIAL_ID` and `CREDENTIAL_USER_LABEL`). I don't see a reason to add the credential type here as the event is narrowed down to OTP types already.
* Update the automated test cases
* Do one manual test to see if the email is sent out. See `tests.md` on how to start a mock email server

Let me know what you think of this approach.

resah commented:
@ahus1 - Thank you, this is a sensible suggestion.

I adjusted the pull request accordingly.

ahus1 reviewed (approved):
Thank you for this change and sorry it took so long. I tested the changes locally and it looks good.

I found that the emails are only sent out when the user's email address is marked as verified, I'll update the docs in a follow-up PR. 

I verified that the emails are sent out as expected. Even if the code doesn't set the user and client in the lines added, they are already sent in the event, so everything is fine here.

ahus1 commented:
Congratulations @resah to this merged PR!

In case you are looking for a topic in a similar area, issue #10114 might catch an eye: It is about adding events for other credential types (for example WebAuthN). Once those events are present, they would require new email templates in the `EmailEventListenerProvider`. 

Let me know if this or another issue raises your interest for another contribution. 

Best, Alexander

ahus1 commented:
Follow-up issue for the documentation changes created: #27620
